keel-bakery-plugin/src/main/kotlin/com/netflix/spinnaker/keel/bakery/constraint/BakeConstraintEvaluator.kt

```diff
@@ -0,0 +1,73 @@
+package com.netflix.spinnaker.keel.bakery.constraint
+
+import com.netflix.frigga.ami.AppVersion
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.Locatable
+import com.netflix.spinnaker.keel.api.RegionSpec
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.artifacts.DebianArtifact
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.bakery.api.BakeConstraint
+import com.netflix.spinnaker.keel.clouddriver.ImageService
+import com.netflix.spinnaker.keel.constraints.ConstraintState
+import com.netflix.spinnaker.keel.constraints.StatefulConstraintEvaluator
+import com.netflix.spinnaker.keel.constraints.SupportedConstraintType
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import com.netflix.spinnaker.kork.dynamicconfig.DynamicConfigService
+import kotlinx.coroutines.runBlocking
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+
+class BakeConstraintEvaluator(
+  override val eventPublisher: ApplicationEventPublisher,
+  private val imageService: ImageService,
+  private val dynamicConfigService: DynamicConfigService,
+  repository: KeelRepository
+) : StatefulConstraintEvaluator<BakeConstraint>(repository) {
+  override val supportedType = SupportedConstraintType<BakeConstraint>("bake")
+
+  override fun canPromote(
+    artifact: DeliveryArtifact,
+    version: String,
+    deliveryConfig: DeliveryConfig,
+    targetEnvironment: Environment,
+    constraint: BakeConstraint,
+    state: ConstraintState
+  ): Boolean {
+    if (artifact !is DebianArtifact) {
+      return true
+    }
+
+    val locatableResources = targetEnvironment
+      .resources
+      .map(Resource<*>::spec)
+      .filterIsInstance<Locatable<*>>()
+
+    val regions: List<String> = locatableResources.flatMap {
+      it.locations.regions.map(RegionSpec::name)
+    }
+
+    return if (regions.isEmpty()) {
+      true
+    } else {
+      log.debug("Searching for baked image for %s in %s", version, regions.joinToString())
+
+      runBlocking {
+        imageService.getLatestNamedImageWithAllRegionsForAppVersion(
```

Comment:
I don't think this method cares which (baseOs, baseLabel) the artifact version was baked onto, which I think is required behavior.

Reply:
Good point. Will fix